File: src/cchess_board/classifier.py

import numpy as np
import cv2
import onnxruntime

from .base_onnx import BaseONNX

from typing import Tuple, List, Union

def center_crop(image: np.ndarray, target_size: Tuple[int, int]) -> np.ndarray:
    """
    Center crop the image to the target size.

    Args:
        image (np.ndarray): The input image.
        target_size (Tuple[int, int]): The desired output size (height, width).

    Returns:
        np.ndarray: The cropped image.
    """
    h, w, _ = image.shape
    target_w, target_h = target_size

    center_x = w // 2
    center_y = h // 2

    start_x = int(center_x - target_w // 2)
    start_y = int(center_y - target_h // 2)

    cropped_image = image[start_y:start_y + target_h, start_x:start_x + target_w]

    return cropped_image


dict_cate_names = {
    'point': '.',
    'other': 'x',
    'red_king': 'K',
    'red_advisor': 'A',
    'red_bishop': 'B',
    'red_knight': 'N',
    'red_rook': 'R',
    'red_cannon': 'C',
    'red_pawn': 'P',
    'black_king': 'k',
    'black_advisor': 'a',
    'black_bishop': 'b',
    'black_knight': 'n',
    'black_rook': 'r',
    'black_cannon': 'c',
    'black_pawn': 'p',
}

class CLASSIFIER_ONNX(BaseONNX):

    label_2_short = dict_cate_names

    classes_labels = list(dict_cate_names.keys())

    def __init__(self, 
                 model_path,
                 # 输入图片大小
                 input_size=(280, 315), # (w, h)
                 # 图片裁剪大小
                 crop_size=(400, 450), # (w, h)
                 ):
        super().__init__(model_path, input_size)

        self.crop_size = crop_size


    def preprocess_image(self, img_bgr: cv2.UMat):

        img_rgb = cv2.cvtColor(img_bgr, cv2.COLOR_BGR2RGB)
        
        if img_rgb.shape[:2] != self.crop_size:
            # 调整图片大小 执行 center crop
            img_rgb = center_crop(img_rgb, self.crop_size) # dst_size = (w, h)

        # resize 到 input_size
        img_rgb = cv2.resize(img_rgb, self.input_size)

        # normalize mean and std
        img = (img_rgb - np.array([ 123.675, 116.28, 103.53])) / np.array([58.395, 57.12, 57.375])

        img = img.astype(np.float32)
        # 转换为浮点型并归一化
        # img = img.astype(np.float32) / 255.0
        
        # 调整维度顺序 (H,W,C) -> (C,H,W)
        img = np.transpose(img, (2, 0, 1))
        
        # 添加 batch 维度
        img = np.expand_dims(img, axis=0)

        return img
    

    def run_inference(self, image: np.ndarray) -> np.ndarray:
        """
        Run inference on the image.

        Args:
            image (np.ndarray): The image to run inference on.

        Returns:
            tuple: A tuple containing the detection results and labels.
        """
        # 运行推理
        outputs, = self.session.run(None, {self.input_name: image})
        
        return outputs

    def pred(self, img_bgr: List[Union[cv2.UMat, str]]) -> Tuple[List[List[str]], List[List[str]], List[List[float]], str]:
        """
        Predict the detection results of the image.

        Args:
            image (cv2.UMat, str): The image to predict.

        Returns:
          
        """
        img_p = self.preprocess_image(img_bgr)
        labels = self.run_inference(img_p)
        # 校验 labels 的 shape
        assert labels.shape[1:] == (90, 16)

        # shape (90, 16)
        first_batch_labels = labels[0]

        # 获取置信度最高的标签
        # list[int]
        label_indexes = np.argmax(first_batch_labels, axis=-1).tolist()

        # 将标签索引转换为标签
        # list[str]
        label_names = [self.classes_labels[index] for index in label_indexes]

        # list[str]
        label_short = [self.label_2_short[name] for name in label_names]

        # 获取置信度, 根据  first_batch_labels 和 label_indexes
        confidence = first_batch_labels[np.arange(first_batch_labels.shape[0]), label_indexes]

        label_names_10x9 = [label_names[i*9:(i+1)*9] for i in range(10)]
        label_short_10x9 = [label_short[i*9:(i+1)*9] for i in range(10)]
        confidence_10x9 = [confidence[i*9:(i+1)*9] for i in range(10)]


        layout_str = "\n".join(["".join(row) for row in label_short_10x9])

        return label_names_10x9, label_short_10x9, confidence_10x9, layout_str
    
    def draw_pred(self, image: cv2.UMat, label_index: int, label_name: str, label_short: str, confidence: float) -> cv2.UMat:

        # 在图像上绘制预测结果
        cv2.putText(image, f"{label_short} {confidence:.2f}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)

        return image
    

    def draw_pred_with_result(self, image: cv2.UMat, results: List[Tuple[int, str, str, float]], cells_xyxy: np.ndarray, is_rgb: bool = True) -> cv2.UMat:

        assert len(results) == cells_xyxy.shape[0]

        if not is_rgb:
            image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)

        for i, (label_index, label_name, label_short, confidence) in enumerate(results):
            # 确保坐标是整数类型
            x1, y1, x2, y2 = map(int, cells_xyxy[i])

            if label_name.startswith('red'):
                color = (180, 105, 255) # 粉红色
            elif label_name.startswith('black'):
                color = (0, 100, 50) # 黑色
            else:
                color = (0, 0, 255) # 蓝色
            
            if confidence < 0.5:
                # yellow 
                color = (255, 255, 0)

            # confidence:.2f 仅保留两位小数 移除

            label_str = f"{label_short} {confidence:.2f}" if confidence < 0.9 else f"{label_short}"

            cv2.putText(image, label_str, (x1 + 8, y2 - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.3, color, 1)

        return image

File: cchess-board-rs/src/classifier.rs
use ndarray::Array4;
use tract_onnx::prelude::tvec;

use crate::base_onnx::{create_model, BaseOnnxError, Image};

/// Mapping from class name to short symbol (FEN-style).
const DICT_CATE_NAMES: [(&str, &str); 16] = [
    ("point", "."),
    ("other", "x"),
    ("red_king", "K"),
    ("red_advisor", "A"),
    ("red_bishop", "B"),
    ("red_knight", "N"),
    ("red_rook", "R"),
    ("red_cannon", "C"),
    ("red_pawn", "P"),
    ("black_king", "k"),
    ("black_advisor", "a"),
    ("black_bishop", "b"),
    ("black_knight", "n"),
    ("black_rook", "r"),
    ("black_cannon", "c"),
    ("black_pawn", "p"),
];

const MEAN: [f32; 3] = [123.675, 116.28, 103.53];
const STD: [f32; 3] = [58.395, 57.12, 57.375];

pub struct ClassifierResult {
    pub label_names: Vec<Vec<String>>,
    pub label_short: Vec<Vec<String>>,
    pub confidence: Vec<Vec<f32>>,
    pub layout_str: String,
}

pub type OnnxModel = tract_onnx::prelude::SimplePlan<
    tract_onnx::prelude::TypedFact,
    Box<dyn tract_onnx::prelude::TypedOp>,
    tract_onnx::prelude::TypedModel,
>;

/// ONNX classifier for Chinese chess pieces on a 10x9 grid.
pub struct ClassifierOnnx {
    model: OnnxModel,
    pub input_size: (usize, usize),
    pub crop_size: (usize, usize),
}

impl ClassifierOnnx {
    pub fn new<P: AsRef<std::path::Path>>(
        model_path: P,
        input_size: (usize, usize),
        crop_size: (usize, usize),
    ) -> Result<Self, BaseOnnxError> {
        let model = create_model(model_path)?;
        let plan = tract_onnx::prelude::SimplePlan::new(model)
            .map_err(|e| BaseOnnxError::OnnxError(e.to_string()))?;
        Ok(Self {
            model: plan,
            input_size,
            crop_size,
        })
    }

    fn center_crop(&self, img: &Image) -> Image {
        let (cw, ch) = self.crop_size;
        let start_x = (img.width / 2).saturating_sub(cw / 2);
        let start_y = (img.height / 2).saturating_sub(ch / 2);
        img.crop(start_x, start_y, cw, ch)
    }

    pub fn preprocess_image(&self, img_bgr: &Image) -> Result<Array4<f32>, BaseOnnxError> {
        let img_rgb = if img_bgr.width != self.crop_size.0 || img_bgr.height != self.crop_size.1 {
            self.center_crop(img_bgr).bgr_to_rgb()
        } else {
            img_bgr.bgr_to_rgb()
        };

        let resized = img_rgb.resize(self.input_size.0, self.input_size.1);

        let (w, h) = (self.input_size.0, self.input_size.1);
        let mut array = Array4::<f32>::zeros((1, 3, h, w));

        for y in 0..h {
            for x in 0..w {
                let pixel = resized.get_pixel(x, y);
                for c in 0..3 {
                    let val = pixel[c] as f32;
                    array[[0, c, y, x]] = (val - MEAN[c]) / STD[c];
                }
            }
        }
        Ok(array)
    }

    pub fn pred(&self, img_bgr: &Image) -> Result<ClassifierResult, BaseOnnxError> {
        use tract_onnx::prelude::Tensor;

        let input = self.preprocess_image(img_bgr)?;
        let shape = input.shape();
        let tensor = Tensor::from_shape(shape, input.as_slice().unwrap())
            .map_err(|e| BaseOnnxError::OnnxError(e.to_string()))?;

        let result = self
            .model
            .run(tvec![tensor.into()])
            .map_err(|e| BaseOnnxError::OnnxError(e.to_string()))?;

        let output = result
            .first()
            .ok_or_else(|| BaseOnnxError::OnnxError("No output from model".into()))?;
        let tensor_f32 = output
            .to_array_view::<f32>()
            .map_err(|e| BaseOnnxError::OnnxError(e.to_string()))?;

        let data: Vec<f32> = tensor_f32.iter().cloned().collect();
        let expected_len = 1 * 90 * 16;
        if data.len() != expected_len {
            return Err(BaseOnnxError::DimMismatch {
                expected: format!("{}", expected_len),
                actual: format!("{}", data.len()),
            });
        }

        let mut label_indexes = vec![0usize; 90];
        let mut confidences = vec![0.0f32; 90];

        for i in 0..90 {
            let mut max_idx = 0;
            let mut max_val = f32::NEG_INFINITY;
            for j in 0..16 {
                let val = data[i * 16 + j];
                if val > max_val {
                    max_val = val;
                    max_idx = j;
                }
            }
            label_indexes[i] = max_idx;
            confidences[i] = max_val;
        }

        let class_keys: Vec<&str> = DICT_CATE_NAMES.iter().map(|(k, _)| *k).collect();
        let label_names: Vec<String> = label_indexes
            .iter()
            .map(|&idx| class_keys[idx].to_string())
            .collect();
        let label_short: Vec<String> = label_indexes
            .iter()
            .map(|&idx| DICT_CATE_NAMES[idx].1.to_string())
            .collect();

        let label_names_10x9: Vec<Vec<String>> = (0..10)
            .map(|i| label_names[i * 9..(i + 1) * 9].to_vec())
            .collect();
        let label_short_10x9: Vec<Vec<String>> = (0..10)
            .map(|i| label_short[i * 9..(i + 1) * 9].to_vec())
            .collect();
        let confidence_10x9: Vec<Vec<f32>> = (0..10)
            .map(|i| confidences[i * 9..(i + 1) * 9].to_vec())
            .collect();

        let layout_str = label_short_10x9
            .iter()
            .map(|row| row.join(""))
            .collect::<Vec<_>>()
            .join("\n");

        Ok(ClassifierResult {
            label_names: label_names_10x9,
            label_short: label_short_10x9,
            confidence: confidence_10x9,
            layout_str,
        })
    }
}
